### ISSUE TITLE
In-Game: [Desktops] Nintendo Switch gamepads connected via USB require filtering from the list of gamepad devices

### ISSUE THREAD

**Issue by offalynne** [YoYoGames/GameMaker-Bugs#8426]
### Description

*Filing issue per Russel's request for LTS2025*

Under normal circumstances, Nintendo Switch controllers do not work on Windows, Mac, or Linux over USB. However a certain HID write puts them in an active mode. Software that does this includes Steam, web browsers, and more. When in active mode, the HID devices blinker wildly which results in persistent random joystick input. In most cases such software creates an additional virtual controller, which is intended to be the device seen/used by game software. To address this, Switch USB devices need to be "blocked" upon gamepad discovery.

Here are GUIDs that always indicate a Nintendo Switch device over USB, and should always be blocked:

```
030000007e0500000e20000000000000 Windows Joy-Con Charger
030000007e0500000e20000000020000 MacOS Joy-Con Charger
030000007e0500000e20000011010000 Linux Joy-Con Charger
030000007e0500000e20000011810000 Linux Joy-Con Charger
030000007e0500000920000000020000 MacOS Switch Pro Controller
030000007e0500000920000011010000 Linux Switch Pro Controller
```

These GUIDs are platform-unique, (other platforms result in different GUIDs for these same devices) so there is no explicit need for this change to be platform-selective.

Unfortunately on Windows there is no GUID differentiation between Bluetooth (usually leads with bus, `0500...`), so a Switch Pro controller can not be blocked on Windows by GUID alone. We have found that if `gamepad_get_button_count()` returns greater than `21`, this indicates a Switch Pro Controller over USB, but not Bluetooth (where it behaves normally). So in addition to the above GUIDs, `030000007e0500000920000000000000` should be blocked if and only-if greater than 21 buttons are found for the device.

In case of a match to any of the above, the gamepad index should behave as if none is connected, so:
1. The `Async - System` gamepad discovery event should not be registered
2. The `Async - System` DS map `async_load`  should not push `"gamepad discovered"` or `"pad_index"` to `async_load`
3. The index should return `false` from `gamepad_is_connected()`
4. The index should return "none" from `gamepad_get_guid()`
5. The index should return "" from `gamepad_get_description()`

If anyone assigned to this lacks a device to test, I suggest testing another DInput device on hand (eg. PlayStation controller) to confirm the implemented blocking method works. Upon follow up I am glad to confirm a working solution locally with a relevant device from the list.

See also #8425, #8426, #8427, #8428

### Steps To Reproduce

Connect gamepad, observe issue

### How reliably can you recreate this issue using your steps above?

Always

### Which version of GameMaker are you reporting this issue for?

2024.8.1 (Monthly)

### Which platform(s) are you seeing the problem on?

Windows

### Sample Package Attached?

- [x] I have attached my bug package

### Sample Project Added?

- [ ] I have included a small sample project

**Comment by rwkay** [comment]
Fix added to internal `develop` version should be in 2024.13 Beta

**Comment by offalynne** [comment]
> *Windows by GUID [...] `030000007e0500000920000000000000` should be blocked if and only-if greater than 21 buttons are found for the device*

Fails, device is not filtered
![Image](https://github.com/user-attachments/assets/b2135886-f256-4ba6-afa3-cbb0fe2d7e4f)

**Comment by rwkay** [comment]
I blocked the one in the message above which is 030000007e0500000e20000000000000, NOTE e2 in the middle - the one in your screenshot it 92 - should I block that one too?

Russell

**Comment by rwkay** [comment]
Added a fix to internal `develop`

**Comment by offalynne** [comment]
Testing v9.9.1.1021

The device is now blocked in both Windows cases regardless of `gamepad_button_count()`
21 or less buttons indicates Bluetooth connectivity, `030000007e0500000e20000000000000` which should *not* be blocked.

**Comment by offalynne** [comment]
A new GUID has been found that indicates a Switch Pro pad in the bad state (over USB), please add `030000007e0500000920000010020000` to the device filter @YYDan @rwkay

**Comment by rwkay** [comment]
> 2 more GUIDs found exhibiting this issue in the wild that need to be blocked as of runtime-2024.13.1.242
> 
> `030000007e0500000920000001000000` `030000007e0500000920000010020000`
> 
> [@YYDan](https://github.com/YYDan) [@cameron-home](https://github.com/cameron-home) [@rwkay](https://github.com/rwkay)

I presume these are just on Windows platform?

**Comment by rwkay** [comment]
Requested GUIDs are now blocked in 2024.14

**Comment by cameron-home** [comment]
As we do not have the devices to test this, I'll close as verified and if the issue still occurs feel free to reopen or log a new bug.

**Comment by SolidplasmaStudios** [comment]
Plugged in a NS Pro Controller with GUID 030000007e0500000920000000000000 and 32 buttons, still got the issue that gives haywire input. Is there any way to support Nintendo Controllers in Game Maker games right now?

**Comment by offalynne** [comment]
@MisutaaAsriel report your issues with Deltarune to the developer of Deltarune. You are wasting the time of 7+ other people with notifications about your unrelated irrelevant problem in an issue that has been closed.

**Comment by SolidplasmaStudios** [comment]
> [@MisutaaAsriel](https://github.com/MisutaaAsriel) report your issues with Deltarune to the developer of Deltarune. You are wasting the time of 7+ other people with notifications about your unrelated irrelevant problem in an issue that has been closed.

This is rather rude and unhelpful considering this has been established as a bug in the engine, and that @cameron-home said we could update if new info about the bug turned up.

**Comment by offalynne** [comment]
@SolidplasmaStudios I triaged this issue with Russel. Cameron's response was to my post. The contents of both yours and the post I replied to demonstrate you do not understand the issue being discussed. Please open a new issue if you have a problem with 2024.14.
